test/extended/util/metrics.go

```diff
@@ -0,0 +1,216 @@
+				return fmt.Errorf("operator pod %q in namespace %q is making %v requests per second. Maximum allowed is %v per second", info.name, info.namespace, info.qps, maxQPSAllowed)
+			}
+		}
+	}
+	return nil
+}
+
+func getPodInfoForNamespace(oc *CLI, namespace string) ([]*podInfo, error) {
+	pods, err := oc.AdminKubeClient().CoreV1().Pods(namespace).List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	podInfos := []*podInfo{}
+
+	for _, pod := range pods.Items {
+		info := &podInfo{
+			name:      pod.Name,
+			namespace: pod.Namespace,
+			status:    string(pod.Status.Phase),
+		}
+
+		if len(pod.Spec.Containers) == 0 {
+			info.result = "skipped, no containers found"
+			info.failed = true
+			podInfos = append(podInfos, info)
+			continue
+		}
+
+		metrics, err := getPodMetrics(oc, &pod)
+		if err != nil {
+			info.result = fmt.Sprintf("error: %s", err)
+			info.failed = true
+			podInfos = append(podInfos, info)
+			continue
+		}
+
+		metricGroup, ok := metrics["rest_client_requests_total"]
+		if !ok {
+			info.result = fmt.Sprintf("error: failed to find counter: %q", "rest_client_requests_total")
+			info.failed = true
+			podInfos = append(podInfos, info)
+			continue
+		}
+
+		procStartTime, ok := metrics["process_start_time_seconds"]
+		if !ok || len(procStartTime.Metric) == 0 {
+			info.result = fmt.Sprintf("error: failed to obtain total cpu time for operator process")
+			info.failed = true
+			podInfos = append(podInfos, info)
+			continue
+		}
+		procStartTimeSeconds := procStartTime.Metric[0].GetGauge().GetValue()
+		totalProcTimeSeconds := time.Now().Unix() - int64(procStartTimeSeconds)
+
+		totalRequestCount := float64(0)
+		for _, metric := range metricGroup.Metric {
+			totalRequestCount += metric.Counter.GetValue()
+		}
+
+		comment := "within QPS bounds"
+		qps := totalRequestCount / float64(totalProcTimeSeconds)
+		if qps > maxQPSAllowed {
+			comment = "exceeds QPS bounds"
+		}
+		info.status = fmt.Sprintf("%s (%s)", info.status, comment)
+		info.qps = qps
+		info.result = fmt.Sprintf("%v requests over a span of %v seconds", totalRequestCount, totalProcTimeSeconds)
+		podInfos = append(podInfos, info)
+	}
+
+	return podInfos, nil
+}
+
+func getPodMetrics(oc *CLI, pod *v1.Pod) (map[string]*dto.MetricFamily, error) {
+	if len(oc.AdminConfig().BearerToken) == 0 {
+		return nil, fmt.Errorf("no Bearer Token found in provided admin config")
+	}
+
+	command := []string{"/bin/curl", "-H", fmt.Sprintf("%s", "Authorization: Bearer "+oc.AdminConfig().BearerToken), "-k", "https://localhost:8443/metrics"}
+	result, err := execCommandInPod(oc, pod, command)
+	if err != nil {
+		return nil, err
+	}
+
+	return parseRawMetrics(result)
+}
+
+func parseRawMetrics(rawMetrics string) (map[string]*dto.MetricFamily, error) {
+	p := expfmt.TextParser{}
+	return p.TextToMetricFamilies(bytes.NewBufferString(rawMetrics))
+}
+
+func execCommandInPod(oc *CLI, pod *v1.Pod, command []string) (string, error) {
```

Comment:
Wait, why you need to get pod's metrics from inside the pod, don't we publish those data in prometheus? What am I missing here?

Reply:
I do this to get operator-specific metrics per operator pod, rather than whatever is exposed in the apiserver's /metrics endpoint